docs/man/man3/FaceTracker.3
.TH "FaceTracker" 3 "Sun Feb 17 2013" "Lockheed Inanimation" \" -*- nroff -*-
.ad l
.nh
.SH NAME
FaceTracker \- 
.PP
Tracks a face as it moves around\&. This class is a wrapper around OpenCV and provides an eassy to use interface for tracking a face between frames\&. The class exposes several face detection tuning parameters\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <facetracker\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBFaceTracker\fP ()"
.br
.RI "\fIDefault constructor uses device 0\&. \fP"
.ti -1c
.RI "\fBFaceTracker\fP (int deviceID)"
.br
.RI "\fIConstructor uses specified device ID for VideoCapture\&. \fP"
.ti -1c
.RI "void \fBResetTracker\fP ()"
.br
.RI "\fIStops tracking of the current face\&. \fP"
.ti -1c
.RI "QRect \fBGetFacePosition\fP ()"
.br
.RI "\fIReturns the rectangle of the currently tracked face\&. \fP"
.ti -1c
.RI "QList< QRect > \fBGetAllFacesPositions\fP ()"
.br
.RI "\fIReturns a list of bounding rectangles of all of the faces\&. \fP"
.ti -1c
.RI "QRect \fBGetBestFacePosition\fP ()"
.br
.RI "\fIReturns the bounding rectangle of the face with the highest confidence factor If there are multiple faces, only the bounding rectangle of the face with the highest confidence is returned\&. \fP"
.ti -1c
.RI "QRect \fBSelectFace2Track\fP (std::vector< cv::Rect > faceRects)"
.br
.RI "\fISelects which face to track from a list Given a list of bounding rectangles, this function select which face will be tracked from here on out\&. \fP"
.ti -1c
.RI "int \fBGetMinFeatureSize\fP ()"
.br
.RI "\fIGetter for minimum feature size used for cv::CascadeClassifier::detectMultiScale() \fP"
.ti -1c
.RI "void \fBSetMinFeatureSize\fP (int minFeatureSize)"
.br
.RI "\fISetter for minimum feature size used for cv::CascadeClassifier::detectMultiScale() \fP"
.ti -1c
.RI "float \fBGetSearchScaleFactor\fP ()"
.br
.RI "\fIGetter for search scale factor used for cv::CascadeClassifier::detectMultiScale() \fP"
.ti -1c
.RI "void \fBSetSearchScaleFactor\fP (float searchScaleFactor)"
.br
.RI "\fISetter for search scale factor used for cv::CascadeClassifier::detectMultiScale() \fP"
.ti -1c
.RI "int \fBGetMinNeighbors\fP ()"
.br
.RI "\fIGetter for minimum neighbors cutoff used for cv::CascadeClassifier::detectMultiScale() \fP"
.ti -1c
.RI "void \fBSetMinNeighbors\fP (int minNeighbors)"
.br
.RI "\fISetter for minimum neighbors cutoff used for cv::CascadeClassifier::detectMultiScale() \fP"
.ti -1c
.RI "void \fBSetProcessingImageDimensions\fP (int width, int height)"
.br
.RI "\fISets the dimensions for the image to be processed by OpenCV This option may effect performance and reliability of the face detection algorithms of OpenCV\&. \fP"
.ti -1c
.RI "unsigned int \fBGetAdditionalFlags\fP ()"
.br
.RI "\fIGetter for flags used for cv::CascadeClassifier::detectMultiScale() These flags are appended to the flags already used for the function call\&. For example, \fBFaceTracker::GetBestFacePosition\fP will use\&. \fP"
.ti -1c
.RI "void \fBSetAdditionalFlag\fP (unsigned int flag)"
.br
.RI "\fIAdds flag to the set of additional flags\&. \fP"
.ti -1c
.RI "void \fBClearAdditionalFlag\fP (unsigned int flag)"
.br
.RI "\fIClears the passed flag from the additional flags\&. \fP"
.ti -1c
.RI "void \fBSetAdditionalFlags\fP (unsigned int flags)"
.br
.RI "\fISets the additional flags to flags\&. \fP"
.ti -1c
.RI "QImage * \fBGetLastImage\fP ()"
.br
.RI "\fIReturns the last processed frame The tracker internally saves the last processed frame and this function can be used to retrieve it\&. Along with the face positions, this can help display real time face tracking information\&. \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBInit\fP (int deviceID)"
.br
.RI "\fIInitialization function\&. \fP"
.ti -1c
.RI "QRect \fBfindClosest\fP (const std::vector< cv::Rect > &rects, QPoint point)"
.br
.RI "\fIFinds rectangle with center closest to point\&. \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "cv::VideoCapture \fBm_vc\fP"
.br
.RI "\fIUsed for acquiring images from camera\&. \fP"
.ti -1c
.RI "cv::CascadeClassifier \fBfaceDetector\fP"
.br
.RI "\fIUsed for face detection\&. \fP"
.ti -1c
.RI "std::string \fBclassifier_xml_filename\fP"
.br
.RI "\fIThe filename of the XML containing the classifier data\&. \fP"
.ti -1c
.RI "QRect \fBm_lastPosition\fP"
.br
.RI "\fIStores the last bounding rectangle of the tracked face\&. \fP"
.ti -1c
.RI "cv::Size \fBm_minFeatureSize\fP"
.br
.RI "\fIMinimum feature size used for cv::CascadeClassifier::detectMultiScale() \fP"
.ti -1c
.RI "float \fBm_searchScaleFactor\fP"
.br
.RI "\fISearch scale factor used for cv::CascadeClassifier::detectMultiScale() \fP"
.ti -1c
.RI "int \fBm_minNeighbors\fP"
.br
.RI "\fIMinimum number of Neighbors to be classified as a face\&. \fP"
.ti -1c
.RI "unsigned int \fBm_additionalFlags\fP"
.br
.RI "\fIAdditional flags passed to cv::CascadeClassifier::detectMultiScale() \fP"
.ti -1c
.RI "unsigned int \fBm_imageWidth\fP"
.br
.RI "\fISpecifies the width of the image on which face detection is performed\&. \fP"
.ti -1c
.RI "unsigned int \fBm_imageHeight\fP"
.br
.RI "\fISpecifies the height of the image on which face detection is performed\&. \fP"
.ti -1c
.RI "cv::Mat \fBcameraFrame_saved\fP"
.br
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static const QRect \fBInvalidQRect\fP"
.br
.RI "\fIEasy way to create an InvalidRect\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
Tracks a face as it moves around\&. This class is a wrapper around OpenCV and provides an eassy to use interface for tracking a face between frames\&. The class exposes several face detection tuning parameters\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "FaceTracker::FaceTracker (intdeviceID)"

.PP
Constructor uses specified device ID for VideoCapture\&. \fBParameters:\fP
.RS 4
\fIdeviceID\fP device to be used for VideoCapture and image acquisition 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "void FaceTracker::ClearAdditionalFlag (unsigned intflag)"

.PP
Clears the passed flag from the additional flags\&. .PP
.nf
m_additionalFlags = m_additionalFlags & (~flag); 
.fi
.PP
 
.PP
\fBParameters:\fP
.RS 4
\fIflag\fP Flag to be cleared 
.RE
.PP

.SS "QRect FaceTracker::findClosest (const std::vector< cv::Rect > &rects, QPointpoint)\fC [private]\fP"

.PP
Finds rectangle with center closest to point\&. \fBParameters:\fP
.RS 4
\fIrects\fP List of rectangles 
.br
\fIpoint\fP Target point 
.RE
.PP

.SS "unsigned int FaceTracker::GetAdditionalFlags ()"

.PP
Getter for flags used for cv::CascadeClassifier::detectMultiScale() These flags are appended to the flags already used for the function call\&. For example, \fBFaceTracker::GetBestFacePosition\fP will use\&. .PP
.nf
CASCADE_FIND_BIGGEST_OBJECT | CASCADE_DO_ROUGH_SEARCH | m_additionalFlags 
.fi
.PP
 for the flags parameter of the cv::CascadeClassifier::detectMultiScale() function call 
.SS "QList<QRect> FaceTracker::GetAllFacesPositions ()"

.PP
Returns a list of bounding rectangles of all of the faces\&. Bounding rectangles of all faces currently in view are returned\&. 
.PP
\fBWarning:\fP
.RS 4
No confidence information is supplied with each face, but all detected faces meet the FaceTracker::minNeighbors factor 
.RE
.PP
\fBReturns:\fP
.RS 4
List of bounding rectangles of all faces 
.RE
.PP

.SS "QRect FaceTracker::GetBestFacePosition ()"

.PP
Returns the bounding rectangle of the face with the highest confidence factor If there are multiple faces, only the bounding rectangle of the face with the highest confidence is returned\&. \fBReturns:\fP
.RS 4
Bounding rectangle of face with highest confidence factor 
.RE
.PP

.SS "QRect FaceTracker::GetFacePosition ()"

.PP
Returns the rectangle of the currently tracked face\&. If there is no face currently being tracked, or this function is called after a call to \fBFaceTracker::ResetTracker\fP, a new face will be selected as per \fBFaceTracker::SelectFace2Track\fP\&.
.PP
\fBReturns:\fP
.RS 4
Bounding rectangle of the face 
.RE
.PP

.SS "QImage * FaceTracker::GetLastImage ()"

.PP
Returns the last processed frame The tracker internally saves the last processed frame and this function can be used to retrieve it\&. Along with the face positions, this can help display real time face tracking information\&. \fBReturns:\fP
.RS 4
Last Processed frame 
.RE
.PP

.SS "void FaceTracker::Init (intdeviceID)\fC [private]\fP"

.PP
Initialization function\&. Performs the initialization of the class, should only be called from constructors 
.PP
\fBParameters:\fP
.RS 4
\fIdeviceID\fP Device to be used for VideoCapture and image acquisition 
.RE
.PP

.SS "void FaceTracker::ResetTracker ()"

.PP
Stops tracking of the current face\&. This Tracker will attempt to follow the same face around as it moves around the field of view of the camera\&. This function stops the tracking of the current face, the next time position information is requested, a new face will be selected to be tracked\&. 
.SS "QRect FaceTracker::SelectFace2Track (std::vector< cv::Rect >faceRects)"

.PP
Selects which face to track from a list Given a list of bounding rectangles, this function select which face will be tracked from here on out\&. \fBParameters:\fP
.RS 4
\fIfaceRects\fP list of faces to select from 
.RE
.PP
\fBReturns:\fP
.RS 4
Bounding rectangle of selected face 
.RE
.PP

.SS "void FaceTracker::SetAdditionalFlag (unsigned intflag)"

.PP
Adds flag to the set of additional flags\&. .PP
.nf
m_additionalFlags = m_additionalFlags | flag; 
.fi
.PP
 
.PP
\fBParameters:\fP
.RS 4
\fIflag\fP Flag to be added to the set of additional flags 
.RE
.PP

.SS "void FaceTracker::SetAdditionalFlags (unsigned intflags)"

.PP
Sets the additional flags to flags\&. .PP
.nf
m_additionalFlags = flags 
.fi
.PP
 
.PP
\fBWarning:\fP
.RS 4
Currently set additional flags are lost! 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIflags\fP The new additional flags 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for Lockheed Inanimation from the source code\&.


src/facetracker.cpp
#include "facetracker.h"
#include <stdexcept>
#include <sstream>
#include <limits>
#include <cmath>
#include <QFile>
#include <QTemporaryFile>
#include <QDebug>
#include <QResource>


const QRect FaceTracker::InvalidQRect(1,1,0,0);

FaceTracker::FaceTracker()
{
    Init(0);
}

FaceTracker::FaceTracker(int deviceID)
{
    Init(deviceID);
}

void FaceTracker::ResetTracker()
{
    //Sets invalid, empty, and null flags for the QRect
    m_lastPosition.setCoords(1,1,0,0);
}

QRect FaceTracker::GetFacePosition(bool normalized)
{
    cv::Mat cameraFrame;
    GetProcessReadyWebcamImage(cameraFrame);
    if(cameraFrame.empty())
        return InvalidQRect;

    std::vector<cv::Rect> faceRects;
    m_faceDetector.detectMultiScale(cameraFrame, faceRects,
                                  m_searchScaleFactor, m_minNeighbors,
                                  cv::CASCADE_SCALE_IMAGE | m_additionalFlags);

    //No Faces detected
    if(faceRects.size() == 0)
        return InvalidQRect;

    //Select new face to track
    if(!m_lastPosition.isValid())
    {
        m_lastPosition = SelectFace2Track(faceRects);
        return m_lastPosition;
    }

    //Find face closest to where the tracked face was last time
    m_lastPosition = findClosest(faceRects,m_lastPosition.center());

    if(normalized)
    {
        return QRect(100*m_lastPosition.x()/m_imageWidth,
                     100*m_lastPosition.y()/m_imageHeight,
                     100*m_lastPosition.width()/m_imageWidth,
                     100*m_lastPosition.height()/m_imageHeight);
    }

    return m_lastPosition;
}

QList<QRect> FaceTracker::GetAllFacesPositions(bool normalized)
{
    cv::Mat cameraFrame;
    GetProcessReadyWebcamImage(cameraFrame);
    if(cameraFrame.empty())
        return QList<QRect>();

    std::vector<cv::Rect> faceRects;
    m_faceDetector.detectMultiScale(cameraFrame, faceRects,
                                  m_searchScaleFactor, m_minNeighbors,
                                  cv::CASCADE_SCALE_IMAGE | m_additionalFlags);
    QList<QRect> qFaceRects;

    if(normalized)
    {
        for(std::vector<cv::Rect>::const_iterator itr = faceRects.begin();
            itr != faceRects.end(); ++itr)
            qFaceRects.append(QRect(itr->x, itr->y, itr->width, itr->height));
    }
    else
    {
        for(std::vector<cv::Rect>::const_iterator itr = faceRects.begin();
            itr != faceRects.end(); ++itr)
            qFaceRects.append(QRect(100*(itr->x)/m_imageWidth,
                                    100*(itr->y)/m_imageHeight,
                                    100*(itr->width)/m_imageWidth,
                                    100*(itr->height)/m_imageHeight));
    }

    return qFaceRects;
}

QRect FaceTracker::GetBestFacePosition(bool normalized)
{
    cv::Mat cameraFrame;
    GetProcessReadyWebcamImage(cameraFrame);
    if(cameraFrame.empty())
        return InvalidQRect;

    std::vector<cv::Rect> faceRects;
    m_faceDetector.detectMultiScale(cameraFrame, faceRects,
                                  m_searchScaleFactor, m_minNeighbors,
                                  cv::CASCADE_FIND_BIGGEST_OBJECT | cv::CASCADE_DO_ROUGH_SEARCH | m_additionalFlags);

    //No Faces detected
    if(faceRects.size() == 0)
        return InvalidQRect;

    if(normalized)
        m_lastPosition.setRect(faceRects[0].x, faceRects[0].y,
                               faceRects[0].width, faceRects[0].height);
    else
        m_lastPosition.setRect(100*faceRects[0].x/m_imageWidth,
                               100*faceRects[0].y/m_imageHeight,
                               100*faceRects[0].width/m_imageWidth,
                               100*faceRects[0].height/m_imageHeight);

    return m_lastPosition;
}

QRect FaceTracker::SelectFace2Track(std::vector<cv::Rect> faceRects)
{
    return findClosest(faceRects, QPoint(m_imageWidth/2, m_imageHeight/2));
}

int FaceTracker::GetMinFeatureSize()
{
    return m_minFeatureSize.height;
}

void FaceTracker::SetMinFeatureSize(int minFeatureSize)
{
    m_minFeatureSize.height = minFeatureSize;
    m_minFeatureSize.width = minFeatureSize;
}

float FaceTracker::GetSearchScaleFactor()
{
    return m_searchScaleFactor;
}

void FaceTracker::SetSearchScaleFactor(float searchScaleFactor)
{
    m_searchScaleFactor = searchScaleFactor;
}

int FaceTracker::GetMinNeighbors()
{
    return m_minNeighbors;
}

void FaceTracker::SetMinNeighbors(int minNeighbors)
{
    m_minNeighbors = minNeighbors;
}

void FaceTracker::SetProcessingImageDimensions(int width, int height)
{
    m_imageWidth = width;
    m_imageHeight = height;

    m_vc.set(CV_CAP_PROP_FRAME_WIDTH, m_imageWidth);
    m_vc.set(CV_CAP_PROP_FRAME_HEIGHT, m_imageHeight);
}

unsigned int FaceTracker::GetAdditionalFlags()
{
    return m_additionalFlags;
}

void FaceTracker::SetAdditionalFlag(unsigned int flag)
{
    m_additionalFlags |= flag;
}

void FaceTracker::ClearAdditionalFlag(unsigned int flag)
{
    m_additionalFlags &= ~flag;
}

void FaceTracker::SetAdditionalFlags(unsigned int flags)
{
    m_additionalFlags = flags;
}

QImage *FaceTracker::GetLastImage()
{
    //! \todo This will corrupt image if function called multiple times, or FaceTracker::GetFaceImage gets called
    cv::cvtColor(m_cameraFrame, m_cameraFrame, CV_BGR2RGB);

    return new QImage(m_cameraFrame.data, m_cameraFrame.cols,
                      m_cameraFrame.rows, QImage::Format_RGB888);
}

QImage *FaceTracker::GetFaceImage()
{
    //! \todo This is very expensive and inefficient way of performing this action
    QImage *image = GetLastImage();
    QImage *result = new QImage(image->copy(m_lastPosition));
    delete image;
    return result;
}


void FaceTracker::Init(int deviceID)
{
    m_minFeatureSize = cv::Size(DEFAULT_MIN_FEATURE_SIZE,DEFAULT_MIN_FEATURE_SIZE);
    m_searchScaleFactor = DEFAULT_SEARCH_SCALE_FACTOR;
    m_minNeighbors = DEFAULT_MIN_NEIGHBORS_CUTOFF;
    m_additionalFlags = DEFAULT_ADDITIONAL_FLAGS;
    m_classifierXmlFilename = DEFAULT_CLASSIFIER_XML_FILENAME;

    m_vc.open(deviceID);
    if(!m_vc.isOpened())
    {
        std::ostringstream error;
        error << "Device at id: " << deviceID << " is not present.";
        throw std::invalid_argument(error.str());
    }

    SetProcessingImageDimensions(DEFAULT_IMAGE_WIDTH, DEFAULT_IMAGE_HEIGHT);

    QResource resource(QString(m_classifierXmlFilename.c_str()));
    if(resource.isValid())
    {
        QFile resFile(QString(m_classifierXmlFilename.c_str()));
        QTemporaryFile *tempFile = QTemporaryFile::createLocalFile(resFile);

        LoadCascadeClassifier(tempFile->fileName().toStdString());

        tempFile->close();
        delete tempFile;
    }
    else
        LoadCascadeClassifier(m_classifierXmlFilename);
}

void FaceTracker::LoadCascadeClassifier(const std::string filename)
{
    try
    {
        m_faceDetector.load(filename);
    }
    catch (...) { }

    if(m_faceDetector.empty())
        throw std::runtime_error("Unable to load classifier xml file.");
}

QRect FaceTracker::findClosest(const std::vector<cv::Rect> &rects, QPoint point)
{
    cv::Rect bestMatch(0,0,0,0);
    float bestDistanceSquared = std::numeric_limits<float>::max();
    for(std::vector<cv::Rect>::const_iterator itr = rects.begin();
        itr != rects.end(); ++itr)
    {
        float dX = fabs((float)((itr->x + (itr->width/2)) - point.x()));
        float dY = fabs((float)((itr->y + (itr->height/2)) - point.y()));

        float distSq = dX*dX + dY*dY;

        if(distSq < bestDistanceSquared)
        {
            bestMatch = *itr;
            bestDistanceSquared = distSq;
        }
    }

    return QRect(bestMatch.x, bestMatch.y, bestMatch.width, bestMatch.height);
}

void FaceTracker::GetProcessReadyWebcamImage(cv::Mat &cameraFrame)
{
    m_vc >> m_cameraFrame;

    if(m_cameraFrame.empty())
    {
        cameraFrame = m_cameraFrame;
        return;
    }

    //To Grayscale
    if(m_cameraFrame.channels() == 3)
        cv::cvtColor(m_cameraFrame, cameraFrame, CV_BGR2GRAY);
    else if(m_cameraFrame.channels() == 4)
        cv::cvtColor(m_cameraFrame, cameraFrame, CV_BGRA2GRAY);

    //Histogram Equalization
    cv::equalizeHist(cameraFrame, cameraFrame);
}
